### PR TITLE
Build on CI with a Fedora 39 image

### PR DESCRIPTION
**Pull request by danyeaw**
This PR changes from using the default Ubuntu runner on GitHub Actions to using Fedora 39. :star_struck:  This allows us to have more recently released versions of GTK and related dependencies to test against, and should unblock #2756 and #2320.

Additionally, I dropped the Python 3.10 build since Python 3.12 is now released.

### PR Checklist
Please check if your PR fulfills the following requirements:

- [X] I have read, and I am following the [Contributor guide](https://github.com/gaphor/gaphor/blob/main/CONTRIBUTING.md)
- [X] I have read, and I understand the GNOME [Code of Conduct](https://wiki.gnome.org/Foundation/CodeOfConduct)

### PR Type
What kind of change does this PR introduce?

- [ ] Bug fix
- [ ] Feature
- [X] Chore (refactoring, formatting, local variables, other cleanup)
- [ ] Documentation content changes

### What is the current behavior?
Linux tests are run on Ubuntu 22.10

Issue Number: N/A

### What is the new behavior?
Linux tests are run on Fedora 39

### Does this PR introduce a breaking change?
- [ ] Yes
- [X] No

### Other information
I initially tried to use a Ubuntu 23.10 image, but I ran in to a lot of problems installing the different versions of Python using the deadsnakes PPA. Ubuntu disables ensurepip and pip wasn't working outside a virtualenv. Fedora was much smoother to get going.

### CODE REVIEW

File: .github/workflows/full-build.yml [hunk 1]
```diff
@@ -50,44 +50,45 @@ jobs:
...
       - name: Install Linux Dependencies
         run: >
-          sudo apt-get update -qq && sudo apt-get install -qq --no-install-recommends upx
-          gir1.2-gtk-4.0 libgirepository1.0-dev libgtksourceview-5-dev libadwaita-1-dev
-          graphviz
-      - name: Set up Python
-        uses: actions/setup-python@65d7f2d534ac1bc67fcd62888c5f4f3d2cb2b236 # v4.7.1
+          dnf install -y gcc git gobject-introspection-devel
+          cairo-gobject-devel graphviz pkg-config gtk4 gtksourceview5-devel
+          libadwaita-devel cairo-devel upx python${{ matrix.python_version }}
```
Comment (amolenaar) [inline .github/workflows/full-build.yml:68]: Is `python${{ matrix.python_version }}` required if `python${{ matrix.python_version }}-devel` is already installed?

File: .github/workflows/full-build.yml [hunk 2]
```diff
@@ -50,44 +50,45 @@ jobs:
...
+          python${{ matrix.python_version }}-devel python-launcher
+          xorg-x11-server-Xvfb
+      - name: Install pip
+        run: python${{ matrix.python_version }} -m ensurepip
```
Comment (amolenaar) [inline .github/workflows/full-build.yml:72]: I would not make this a separate step, but include it with either the previous or next step. Having it as a separate doesn't add anything imho.

File: .github/workflows/full-build.yml [hunk 3]
```diff
@@ -50,44 +50,45 @@ jobs:
...
+        run: |
+          py -${{ matrix.python_version }} -m pip install pipx
+          echo "/github/home/.local/bin" >> $GITHUB_PATH
+      - uses: actions/checkout@8ade135a41bc03ea155e62e844d188df1ea18608 # v4.1.0
         with:
-          python-version: ${{ matrix.python_version }}
-          allow-prereleases: true
+          ref: ${{ github.event.pull_request.head.sha }}
+      - name: Set ownership of Checkout directory
+        run: chown -R $(id -u):$(id -g) $PWD
```
Comment (amolenaar) [inline .github/workflows/full-build.yml:81]: Same here. I'd probably add it to the `Install Linux Dependencies` step.

**Review by amolenaar — APPROVED** [review]
Quite a straighforward change. Nicely done :).

I'd reduce the number of steps though, so we only see the relevant (higher level) ones, such as install and testing.

File: .github/workflows/full-build.yml (hunk 2 above)
Comment (danyeaw) [reply .github/workflows/full-build.yml:72]: Yup, great feedback, thanks!

File: .github/workflows/full-build.yml (hunk 1 above)
Comment (danyeaw) [reply .github/workflows/full-build.yml:68]: Thanks, yup, this simplifies the dependencies a bit.

File: .github/workflows/full-build.yml (hunk 3 above)
Comment (danyeaw) [reply .github/workflows/full-build.yml:81]: In the Linux dependencies step we need to install Git for the checkout dependency to work properly, and then finally set the correct permissions. So we need these 3 steps separately to keep this order.

**Pull request merged**